csi/controller_server.go

```diff
@@ -545,6 +547,48 @@ func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
 		return nil, status.Error(codes.InvalidArgument, "Snapshot name must be provided")
 	}
 
+	volume, err := cs.apiClient.Volume.ById(csiVolumeName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get volume %v", csiVolumeName)
+	}
+	if volume.State == string(longhorn.VolumeStateDetached) {
+		logrus.Debugf("CSI snapshot attempting to attach %v volume to a random node", volume.Name)
+		nodeID, err := cs.autoAttachVolumeToRandomNode(volume)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		logrus.Debugf("CSI snapshot attached %v volume to %v node", volume.Name, nodeID)
+
+		defer func() {
+			volume, err := cs.apiClient.Volume.ById(csiVolumeName)
+			if err != nil {
+				logrus.WithError(err).Errorf("volume %v unavailable", csiVolumeName)
+				return
+			}
+			if volume == nil {
+				logrus.WithError(err).Errorf("volume %v not found", csiVolumeName)
+				return
+			}
+			logrus.Debugf("CSI snapshot attempting to detach %v volume on %s node", csiVolumeName, nodeID)
+			_, err = cs.unpublishVolume(volume, nodeID, func() error {
+				isSharedVolume := requiresSharedAccess(volume, nil) && !volume.Migratable
```

Comment:
Why if the volume is shared volume, we don't need to wait until it is unpublished? Is it because we just assume probably it still used by other workloads? 

If that is the case, you will only do auto-attach only when the volume is detached, so why we need to have this condition to bypass the waiting?

Reply:
Nice catch. BTW, the RWX volume will keep attached when there is **at least one workload** using it. Otherwise, we may still need to consider auto-attach/detach it.
https://github.com/longhorn/longhorn-manager/blob/24b38e1aa25c7522a39dda60cb992793b8136240/controller/share_manager_controller.go#L431-L434

Reply:
Fixed!